README.md
# Backup Secretary - a discord bot

## Getting started

### Project tools:

- This project is made with NodeJS.
- Although Nodemon isn't required, it can save you tons of time from restarting your server manually.

### Starting

- In project's folder, run `npm start`
  - With nodemon: `nodemon`
    - If the folder structure is changed, make sure to update the new location of the index.js on package.json file.

### Sharing

- To add this bot to your server, follow this link: https://discord.com/api/oauth2/authorize?client_id=DISCORD_APPLICATION_CLIENT_ID&scope=bot

### Discord

- Make sure to head over to https://discord.com/developers/applications and get your API key

## Environment

- Add these variables to your "bash profile"

- `BOT_TOKEN` and assign the discord API key
- `GITHUB_WEBHOOK_ID` and assign discord's webhook ID that is also assigned to github
- `GITHUB_WEBHOOK_TOKEN` assign the webhook's token ""

# Join!

- Join my test server! https://discord.gg/gBPYWQs


CHANGELOG.md
#### 0.6.0 (2021-01-19)

##### Chores

*  add help or commands command ([45f97591](https://github.com/EntOfMD/discord-bot/commit/45f97591c52a6db5d48cffc2133817429e055ff5))

##### Documentation Changes

*  add codeowners ([c05ade15](https://github.com/EntOfMD/discord-bot/commit/c05ade1533d5c7192922b6a1beba4e3280ffcfda))

##### New Features

*  add mkdocs with custom theme ([1d432548](https://github.com/EntOfMD/discord-bot/commit/1d432548ed7f9b94945d76baf7402576d442a6a7))
*  add mkdocs with custom theme ([06e1313e](https://github.com/EntOfMD/discord-bot/commit/06e1313e6f9bcb935cbfdbab37f2599098b5633d))

##### Bug Fixes

*  fix GA multiple versions ([e0537e14](https://github.com/EntOfMD/discord-bot/commit/e0537e1498641a57290c614e1b4c3beb5acdd9d5))

##### Refactors

*  refactor the constants ([340ec046](https://github.com/EntOfMD/discord-bot/commit/340ec046347c95b70dd5dccc9103e0e906434025))
*  better handle commands ([b4b49a3d](https://github.com/EntOfMD/discord-bot/commit/b4b49a3d66f335dd7a1f67314a6058523b1edf6b))



src/index.js
const { Client, WebhookClient, MessageEmbed } = require("discord.js");
const { Functions } = require("./util/functions");
const { Constants, Greetings, Commands } = require("./util/constants");

// discord.js uses cached set of data,
// beginning from when the bot was last online.
// use partials to access data that isn't cached.
const client = new Client({
	partials: ["MESSAGE", "REACTION"],
});
const webhookClient = new WebhookClient(
	Constants.env.GITHUB_WEBHOOK_ID,
	Constants.env.GITHUB_WEBHOOK_TOKEN
);
const PREFIX = "%";

// login
client.login(Constants.env.BOT_TOKEN);
client.on("ready", () => {
	client.user
		.setActivity("%help", {
			type: "LISTENING",
			url: "https://entofmd.io",
		})
		.then((presence) => {
			if (presence.status === "online") {
				console.log(`${client.user.tag} is ready`);
			}
		})
		.catch(console.error);
});

// message event listener
client.on("message", (message) => {
	const CONTENT = message.content;

	console.log(`[${message.author.tag}: ${CONTENT}]`);

	// mitigating counting bot's own reply
	// to avoid a loop
	if (message.author.bot) return;

	// checking a list of greetings
	// to see if the greet is recognized
	if (Greetings.includes(CONTENT.toLowerCase())) {
		message.reply(`Hello there!`);
		message.react("👋");
	}

	// commands
	if (CONTENT.startsWith(PREFIX)) {
		const [CMD_NAME, ...args] = CONTENT.trim()
			.substring(PREFIX.length)
			.split(/\s+/);

		console.log(message.author.tag, "used command:", CMD_NAME, args);
		/* TODO
        handle channel vs DM
         - the bot gets events from channel and DM all the same.
         - DM doesn't have channel methods
         - channel doesn't have DM methods
         - currently the bot would crash if someone were to DM a command
          - because 'member' is searching cached server user listing
         */
		if (Commands.includes(CMD_NAME.toLowerCase())) {
			if (message.channel.type == "text") {
				//this is for channel messages
				const MEMBER = message.guild.members.cache.get(args[0]);

				switch (CMD_NAME.toLowerCase()) {
					case "kick":
						Functions.checkKickable(message, MEMBER);
						message.react("☑️");
						break;
					case "ban":
						Functions.checkBannable(message, MEMBER);
						message.react("☑️");
						break;
					case "say":
						message.channel.send(args.join(" "));
						message.react("☑️");
						break;
					case "announce":
						webhookClient.send(args.join(" "));
						message.react("☑️");
						break;
					case "commands":
					case "help":
						Functions.listCommands(message);
						message.react("☑️");
						break;
					default:
						console.log("command received, could not parse command");
						return;
				}
			} else if (message.channel.type == "dm") {
				// this is for DMs
			}
		} else if (Commands.includes(CMD_NAME)) {
			message.reply("invalid arugments");
		} else {
			// message.reply("invalid command");
			Functions.listCommands(message);
		}
	}
});

// reaction event listener
// ADD role
client.on("messageReactionAdd", (reaction, user) => {
	const { name } = reaction.emoji;
	const member = reaction.message.guild.members.cache.get(user.id);

	if (reaction.message.id === Constants.threads.PROGRAMMER_ROLE) {
		switch (name) {
			case "🐉":
				member.roles.add(Constants.roles.GO);
				console.log("Added Go role to", user.tag);
				break;
			case "🍌":
				member.roles.add(Constants.roles.JS);
				console.log("Added JS role to", user.tag);
				break;
			default:
				break;
		}
	}
});

// reaction event listener
// REMOVE role
client.on("messageReactionRemove", (reaction, user) => {
	const { name } = reaction.emoji;
	const member = reaction.message.guild.members.cache.get(user.id);

	if (reaction.message.id === Constants.threads.PROGRAMMER_ROLE) {
		switch (name) {
			case "🐉":
				member.roles.remove(Constants.roles.GO);
				console.log("Removed Go role from", user.tag);
				break;
			case "🍌":
				member.roles.remove(Constants.roles.JS);
				console.log("Removed JS role from", user.tag);
				break;
			default:
				break;
		}
	}
});

// mentioning channel name
// client.on("message", (message) => {
// 	if (!message.content.startsWith(PREFIX)) return;

// 	let channelName = message.channel.name;
// 	const [CMD_NAME, ...args] = message.content
// 		.trim()
// 		.substring(PREFIX.length)
// 		.split(/[^0-9a-zA-Z]|[\/s\\]/);

// 	if (channelName === "bot-dev") {
// 		if (
// 			message.mentions.channels.array().length > 0 &&
// 			args.includes(message.channel.id)
// 		) {
// 			channel_id = message.mentions.channels.first().id;
// 			console.log(CMD_NAME, args);

// 			message.reply("Hi!");
// 		} else {
// 			const channelError = new MessageEmbed()
// 				.setTitle("Huh? Oh!")
// 				.setDescription(
// 					"Hey there! You forgot to mention a channel! You don't know how?!? Do `#channelname`!"
// 				);
// 			message.channel.send(channelError);
// 		}
// 	}
// });

// When a new member joins
client.on("guildMemberAdd", (member) => {
	// Send the message to a designated channel on a server:
	const channel = member.guild.channels.cache.find(
		(ch) => ch.name === "bot-dev"
	);
	// Do nothing if the channel wasn't found on this server
	if (!channel) return;
	// Send the message, mentioning the member
	channel.send(`Welcome to the server, ${member}`);
});

client.on("disconnect", () => {
	console.log("disconnected");
});


src/util/constants.js
const yargs = require("yargs");

const argv = yargs
	.env("BOT_TOKEN")
	.option("BOT_TOKEN", { alias: "t" })
	.option("GITHUB_WEBHOOK_ID", { alias: "ghwhid" })
	.option("GITHUB_WEBHOOK_TOKEN", { alias: "ghwhtk" }).argv;

const Constants = {
	env: {
		BOT_TOKEN: argv.BOT_TOKEN || process.env.BOT_TOKEN,
		GITHUB_WEBHOOK_ID: argv.GITHUB_WEBHOOK_ID || process.env.GITHUB_WEBHOOK_ID,
		GITHUB_WEBHOOK_TOKEN:
			argv.GITHUB_WEBHOOK_TOKEN || process.env.GITHUB_WEBHOOK_TOKEN,
	},
	threads: {
		PROGRAMMER_ROLE: "749711465665003732",
	},
	roles: {
		GO: "749704008876294229",
		JS: "749704270927888425",
	},
};

const Commands = ["kick", "ban", "say", "announce", "commands", "help"];

const Greetings = [
	"g'day",
	"hello",
	"hi",
	"hey",
	"hiya",
	"howdy",
	"sup",
	"wassup",
	"yo",
];

module.exports = { Constants, Commands, Greetings };


src/util/functions.js
const { Commands } = require("./constants");

const Functions = {
	// checks if the user is kickable and
	// kick them async
	checkKickable: (message, member) => {
		if (member) {
			member
				.kick()
				.then((member) => message.channel.send(`kicked ${member}. Adios!`))
				.catch((err) => {
					message.reply(
						"I don't have authority to do that. Their power level is over 9000!"
					);
					throw err;
				});
		} else {
			message.reply("couldn't find that user");
		}
	},
	// checks if the user is bannable and
	// ban them async
	checkBannable: (message, member) => {
		if (member) {
			member
				.ban()
				.then((member) =>
					message.channel.send(`banned ${member}. good riddance`)
				)
				.catch((err) => {
					message.reply(
						"I don't have authority to do that. Their power level is over 9000!"
					);
					throw err;
				});
		} else {
			message.reply("couldn't find that user");
		}
	},
	// returns a list of commands
	listCommands: (message) => {
		if (message) {
			return message.reply("Available commands:\n" + Commands);
		}
	},
	splitLines: (s) => {
		return s.split(/\r\n|\r|\n/);
	},
};

module.exports = { Functions };


src/util/helper.js
const yargs = require("yargs");

const argv = yargs
	.env("BOT_TOKEN")
	.option("BOT_TOKEN", { alias: "t" })
	.option("GITHUB_WEBHOOK_ID", { alias: "ghwhid" })
	.option("GITHUB_WEBHOOK_TOKEN", { alias: "ghwhtk" }).argv;

const Constants = {
	env: {
		BOT_TOKEN: argv.BOT_TOKEN || process.env.BOT_TOKEN,
		GITHUB_WEBHOOK_ID: argv.GITHUB_WEBHOOK_ID || process.env.GITHUB_WEBHOOK_ID,
		GITHUB_WEBHOOK_TOKEN:
			argv.GITHUB_WEBHOOK_TOKEN || process.env.GITHUB_WEBHOOK_TOKEN,
	},
	threads: {
		PROGRAMMER_ROLE: "749711465665003732",
	},
	roles: {
		GO: "749704008876294229",
		JS: "749704270927888425",
	},
};

const Greetings = [
	"g'day",
	"hello",
	"hi",
	"hey",
	"hiya",
	"howdy",
	"sup",
	"wassup",
	"yo",
];

const Commands = ["kick", "ban", "say", "announce", "commands", "help"];

const Functions = {
	// checks if the user is kickable and
	// kick them async
	checkKickable: (message, member) => {
		if (member) {
			member
				.kick()
				.then((member) => message.channel.send(`kicked ${member}. Adios!`))
				.catch((err) => {
					message.reply(
						"I don't have authority to do that. Their power level is over 9000!"
					);
					throw err;
				});
		} else {
			message.reply("couldn't find that user");
		}
	},
	// checks if the user is bannable and
	// ban them async
	checkBannable: (message, member) => {
		if (member) {
			member
				.ban()
				.then((member) =>
					message.channel.send(`banned ${member}. good riddance`)
				)
				.catch((err) => {
					message.reply(
						"I don't have authority to do that. Their power level is over 9000!"
					);
					throw err;
				});
		} else {
			message.reply("couldn't find that user");
		}
	},
	// returns a list of commands
	listCommands: (message) => {
		if (message) {
			test = [``];
			return message.reply("Available commands:\n" + Commands);
		}
	},
};

module.exports = { Greetings, Commands, Functions, Constants };
